barman/server.py

```diff
@@ -265,9 +269,20 @@ def __init__(self, config):
                     raise ConninfoException(
                         "Missing 'conninfo' parameter for server '%s'" % config.name
                     )
-                self.postgres = PostgreSQLConnection(
-                    config.conninfo, config.immediate_checkpoint, config.slot_name
-                )
+                # If primary_conninfo is set then we're connecting to a standby
+                if config.primary_conninfo is not None:
+                    self.postgres = StandbyPostgreSQLConnection(
+                        config.conninfo,
+                        config.immediate_checkpoint,
+                        config.slot_name,
+                        # The standby needs a reference to the primary so that it can
+                        # perform WAL switches itself when calling pg_backup_stop.
+                        primary_conninfo=config.primary_conninfo,
+                    )
+                else:
+                    self.postgres = PostgreSQLConnection(
```

Comment:
If a user has not set `primary_conninfo` it is still possible that `PostgreSQLConnection` is a standby server - in this case Barman will do nothing special about it, it will attempt to switch WALs on the server at the usual points and ignore the `PostgresInRecovery` exception from psycopg2 just like it does now.

The user will therefore see the same behaviour as with the current Barman, unless they set `primary_conninfo`.